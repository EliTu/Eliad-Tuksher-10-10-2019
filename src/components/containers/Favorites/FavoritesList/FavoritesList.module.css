@import '../../../../variables.css';

.FavoritesListStyles {
    width: 100%;
    display: grid;
    grid-template-columns: 25% 25% 25% 25%;
    grid-column-gap: 0.7rem;
    /* grid-column: auto; */

}

.CardWrapper {
    width: 100%;
    cursor: pointer;
    margin: 0.8 auto;
    transition: all 0.15s ease-in-out;
    position: relative;
}

.CardWrapper:hover {
    transform: scale(1.1);
    box-shadow: var(--shadow-light-spread);
}

.CardWrapper div {
    width: 90%;
    margin: 0.8rem auto;
}

.removeButtonStyle {
    z-index: 2;
    position: absolute;
    top: 10%;
    right: 10%;
    border: block;
    background-color: rgba(206, 202, 202, 0.932);
    width: 10%;
    height: 2rem;
    border-radius: 50%;
    display: block;
    font-size: 1.45rem;
    font-weight: 800;
    color: rgba(233, 229, 229, 0.932);

}