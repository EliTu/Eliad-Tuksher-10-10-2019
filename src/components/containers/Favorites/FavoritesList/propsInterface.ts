export interface IProps {
	favorites: { key: string; cityName: string; countryName: string }[];
	favoritesWeatherData: {
		key: string;
		cityName: string;
		countryName: string;
		WeatherText: string;
		Temperature: { Metric: { Value: number; Unit: string } };
	}[];
	getFavoritesWeatherData: (
		key: string,
		countryName: string,
		cityName: string,
		dispatchIdentifier: string
	) => void;
	history: any;
	getClickedItemWeatherCurrentData: (
		key: string,
		countryName: string,
		cityName: string,
		dispatchIdentifier: string
	) => void;
	getClickedItemFiveDaysForecast: (key: string) => void;
}
