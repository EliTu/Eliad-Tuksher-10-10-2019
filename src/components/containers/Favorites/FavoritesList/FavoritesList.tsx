import React, { useEffect, memo } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import Card from '../../../display/UI/Card/Card';
import { fireCurrentWeatherHttpRequest } from '../../WeatherDisplayContainer/store/actions';
import { fireFiveDaysForecastHttpRequest } from '../../WeatherDisplayContainer/store/actions';
import styles from './FavoritesList.module.css';
import { IProps } from './propsInterface';

export const FavoritesList: React.FC<IProps> = ({
	favorites,
	favoritesWeatherData,
	getFavoritesWeatherData,
	history,
	getClickedItemWeatherCurrentData,
	getClickedItemFiveDaysForecast,
}) => {
	const { FavoritesListStyles, CardWrapper } = styles;

	useEffect(() => {
		favorites.map(({ key, cityName, countryName }) =>
			getFavoritesWeatherData(
				key,
				cityName,
				countryName,
				'favortiesWeather'
			)
		);
	}, [favorites, getFavoritesWeatherData]);

	const handleFavoriteCardClick = (
		key: string,
		cityName: string,
		countryName: string
	) => {
		getClickedItemWeatherCurrentData(
			key,
			cityName,
			countryName,
			'currentWeather'
		);
		getClickedItemFiveDaysForecast(key);
		history.push('/');
	};

	return (
		<div className={FavoritesListStyles}>
			{favoritesWeatherData.length > 0 ? (
				favoritesWeatherData.map(
					({
						cityName,
						countryName,
						key,
						WeatherText,
						Temperature,
					}) => {
						const { Metric } = Temperature;

						return (
							<div
								className={CardWrapper}
								key={key}
								onClick={() =>
									handleFavoriteCardClick(
										key,
										cityName,
										countryName
									)
								}
							>
								<Card
									mainHeading={`${cityName}, ${countryName}`}
									description={WeatherText}
									date={'Current:'}
									temperatures={`${Metric.Value}${Metric.Unit}`}
								/>
							</div>
						);
					}
				)
			) : (
				<p>
					There seems to be no favorite items! Please search for your
					favorite destination and press on the
					<span role="img"> ‚ù§ </span>
					button to add to your favorites
				</p>
			)}
		</div>
	);
};
// Redux setup:
const mapStateToProps = (state: any) => {
	return {
		favorites: state.favorites.favoritesList,
		favoritesWeatherData: state.favorites.weatherData,
	};
};

const mapDispatchToProps = (dispatch: any) => {
	return {
		getFavoritesWeatherData: (
			key: string,
			cityName: string,
			countryName: string
		) =>
			dispatch(
				fireCurrentWeatherHttpRequest(
					key,
					'favoritesWeather',
					cityName,
					countryName
				)
			),
		getClickedItemWeatherCurrentData: (
			key: string,
			cityName: string,
			countryName: string
		) =>
			dispatch(
				fireCurrentWeatherHttpRequest(
					key,
					'currentWeather',
					cityName,
					countryName
				)
			),
		getClickedItemFiveDaysForecast: (key: string) =>
			dispatch(fireFiveDaysForecastHttpRequest(key)),
	};
};

export default withRouter(
	connect(
		mapStateToProps,
		mapDispatchToProps
	)(memo(FavoritesList))
);
