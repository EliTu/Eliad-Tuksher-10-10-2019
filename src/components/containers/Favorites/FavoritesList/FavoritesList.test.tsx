import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { FavoritesList } from './FavoritesList';
import initialShallowRender from '../../../../utilities/test-utilities/initialShallowRender';

describe('FavoritesList componenet', () => {
	let component: ShallowWrapper;
	const currentWeatherCallback = jest.fn();
	const fiveDaysForecastCallback = jest.fn();
	beforeEach(
		() =>
			(component = shallow(
				<FavoritesList
					favorites={[
						{
							key: '123',
							cityName: 'munich',
							countryName: 'germany',
						},
					]}
					favoritesWeatherData={[
						{
							key: '123',
							cityName: 'munich',
							countryName: 'germany',
							WeatherText: 'warm',
							Temperature: { Metric: { Value: 23, Unit: 'c' } },
						},
						{
							key: '456',
							cityName: 'germering',
							countryName: 'germany',
							WeatherText: 'warm',
							Temperature: { Metric: { Value: 24, Unit: 'c' } },
						},
					]}
					history={{ push: () => {} }}
					getFavoritesWeatherData={() => {}}
					getClickedItemWeatherCurrentData={currentWeatherCallback}
					getClickedItemFiveDaysForecast={fiveDaysForecastCallback}
				/>
			))
	);

	it('should render without errors', () => {
		initialShallowRender(component, 'FavoritesListStyles');
	});

	it('should render a div with the relevant favorite weather data', () => {
		expect(component.children().find('div')).toExist();
		expect(component.children().find('div').length).toBe(2);
	});
});
