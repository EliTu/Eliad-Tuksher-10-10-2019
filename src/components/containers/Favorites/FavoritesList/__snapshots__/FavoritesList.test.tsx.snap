// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesList componenet should display a message if there are no favorites to display 1`] = `
<div
  className="FavoritesListStyles"
/>
`;

exports[`FavoritesList componenet should render without errors 1`] = `
<div
  className="FavoritesListStyles"
>
  <div
    className="CardWrapper"
    key="123"
    onClick={[Function]}
  >
    <Card
      description="warm"
      mainHeading="munich, germany"
      temperatures="23"
    />
  </div>
  <div
    className="CardWrapper"
    key="456"
    onClick={[Function]}
  >
    <Card
      description="warm"
      mainHeading="germering, germany"
      temperatures="24"
    />
  </div>
</div>
`;
